notebooks/torch_compat.py

```diff
@@ -0,0 +1,154 @@
+
+def all_mag(x, EPS=1e-7):
+    if x.shape[-1] == 2:
+        power = x.pow(2).sum(-1) + EPS
+        return power.pow(0.5)
+    return mag(x, -2, EPS=EPS)
+
+
+def plot(*specs):
+    for s in specs:
+        mag_spec = all_mag(s).data.numpy()
+        plt.figure()
+        plt.imshow(mag_spec[0].T)
+        plt.colorbar()
+        plt.show(block=False)
+
+
+def to_asteroid(x):
+    return torch.cat([x[..., 0], x[..., 1]], dim=-2)
+
+
+def square_ola(window: torch.Tensor, kernel_size: int, stride: int, n_frame: int) -> torch.Tensor:
+    window_sq = window.pow(2).view(1, -1, 1).repeat(1, 1, n_frame)
+    return torch.nn.functional.fold(
+        window_sq, (1, (n_frame - 1) * stride + kernel_size), (1, kernel_size), stride=(1, stride)
+    ).squeeze(2)
+
+
+@script_if_tracing
+def pad_all_shapes(x: torch.Tensor, pad_shape: List[int], mode: str = "reflect") -> torch.Tensor:
+    if x.ndim < 3:
+        return F.pad(x[None, None], pad=pad_shape, mode=mode).squeeze(0).squeeze(0)
+    return F.pad(x, pad=pad_shape, mode=mode)
+
+
+class TorchSTFT(STFTFB):
+    def __init__(self, *args, center=True, normalize=False, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.center = center
+        self.normalize = normalize
+
+    def pre_analysis(self, wav):
+        if not self.center:
+            return wav
+        pad_shape = [self.kernel_size // 2, self.kernel_size // 2]
+        wav = pad_all_shapes(wav, pad_shape=pad_shape, mode="reflect")
+        return wav
+
+    def post_analysis(self, spec):
+        spec[..., 0, :] *= np.sqrt(2)
+        spec[..., self.n_filters // 2, :] *= np.sqrt(2)
+        return spec * (self.kernel_size // 2) ** 0.5
+
+    def pre_synthesis(self, spec):
+        spec = spec.clone()
+        spec[..., 0, :] /= np.sqrt(2)
+        spec[..., self.n_filters // 2, :] /= np.sqrt(2)
+        return spec / (self.kernel_size // 2) ** 0.5
+
+    def post_synthesis(self, wav):
+        # Assume the window is the same in the STFT and ISTFT
+        # OLA the squared windows, detect zeros and divide by it.
+        # In pytorch this is the error.
+        # RuntimeError: istft(torch.FloatTensor[129, 124, 2], n_fft=256,
+        # hop_length=128, win_length=256, window=torch.FloatTensor{[256]},
+        # center=0, normalized=0, onesided=1, length=16000)window overlap add min: 0
+
+        n_frame = 1 + (wav.shape[-1] - self.kernel_size) // self.stride
+        wsq_ola = square_ola(
+            torch.from_numpy(self.torch_window),
+            kernel_size=self.kernel_size,
+            stride=self.stride,
+            n_frame=n_frame,
+        )
+        view_shape = [1 for _ in wav.shape[:-1]] + [-1]
+        wsq_ola = wsq_ola.view(view_shape)
+
+        start = self.kernel_size // 2 if self.center else 0
+        wav = wav[..., start:]
+        wsq_ola = wsq_ola[..., start:]
+
+        min_mask = wsq_ola.abs() < 1e-11
+        if min_mask.any():
+            # Warning instead of error. Might be trimmed afterward.
```

Comment:
the trimming is done only if length is used correct ?

Reply:
In `torch.istft`

> The :attr:`n_fft`, :attr:`hop_length`, :attr:`win_length` are all the same which prevents the calculation
of right padding. These additional values could be zeros or a reflection of the signal so providing
:attr:`length` could be useful. If :attr:`length` is ``None`` then padding will be aggressively removed
(some loss of signal).

So, if there is no `length`, the output signal is always smaller than the input, which is weird for me.

Reply:
even of the signal has a length of `% win_lenght == 0` ?

Reply:
No, indeed. In this case, the signal is the same.